### ISSUE TITLE
Complete swipe gesture support for all platforms if possible?

### ISSUE THREAD

**Issue by houser42** [Nerwyn/android-tv-card#13]
I think complete support for all swipe gestures that are available on the  respective platforms would be a very valuable addition to this project. I have no idea what the magnitude of work would be, but I hope it is already there in the webkits and can be added. Perhaps it needs to be added natively to the HA apps?

I am mainly a Mac and Ubuntu user and these gestures are there for general use of websites both on
iPhone, iPad and on MacOS and Ubuntu touchpads when using Safari, Chrome and Firefox. 
I use Windows to some extent and it appears very similar and is of course a ubiquitous feature on millions of handsets worldwide.
Complete support for this would surely be a very useful and familiar feature for many, many users?
 
Additional gestures that I would find useful in order of priority for my use for remotes on the HA App on iPhones as well as MacOS and also if possible mainstream browsers on desktop computer touchpads for relevant platforms :
1. one-finger double-tap
3. two finger swipe up/down/left right  (to swipe left-right-up-down on HA app and most browsers)
4. two finger-tap (as in control-click on most browsers)
5. two finger double-tap
6. fast two-finger swipe for acceleration of scrolling in lists horizontal and vertical
7. long-press (already works great as center long-press now)

**Comment by Nerwyn** [comment]
Haven't forgotten about this, but wanted to get the TypeScript refactor done first and afterwards the [Kodi keyboard](https://github.com/Nerwyn/android-tv-card/issues/15) seemed like a relatively lower hanging fruit to implement.

#### Implementation notes:
> 1. One finger double tap

Already implemented but doesn't work on iPhone due to limitations in iOS webview. Searching for "dblclick event on iphone" shows me a lot of people with the same issue but not many solutions.

> 2. Two finger swipes

I think I can implement this using [TouchEvent touches or targetTouches](https://developer.mozilla.org/en-US/docs/Web/API/TouchEvent/targetTouches).

> 3. Two finger tap

This may be a bit more tricky as single and double tap events are actually [MouseEvents](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent). I could look into switching touchpad taps to touch events instead and then I could possibly use the touch event touches to determine how many fingers are being used.

> 4. Two finger double tap.

Same as above. Can possibly be implemented by switching the touchpad to use touch events on regular taps but the iOS double click issue still remains. Although this makes me wonder if switching the touchpad to touch events for clicks would help fix that?

> 5. Fast two finger swipes for scrolling.

I'm not sure if there is a path forward for this one. The Android TV remote doesn't have a scroll command. It may be possible to send it [using ADB](https://stackoverflow.com/questions/39190083/how-can-i-scroll-an-application-using-adb) instead but from what I've seen with the keyboard ADB commands have higher latency.

> 6. Long press

This should already be implemented as the `long_click_keycode` config option to remap lock center clicks to any other key.

**Comment by houser42** [comment]
Appreciate the heads-up, sounds good!  Many thanks.
Please also if possible, keep in mind enabling swipe gestures for any browser on any platform as is normal behaviour.
Using these gestures in say Chrome on an Android device or Edge or Firefox on any platform etc.
I would assume that if possible this should all be generic and platform agnostic in terms of end user experience, even though some platforms may need special tweaks to work as the others. Something like that?

**Comment by Nerwyn** [comment]
Unfortunately touchpad support on all platforms is conditional on them supporting touch events. While it works on most major browsers, touch events only trigger if a touch screen is used. Taps on the touchpad work because browsers convert mouse clicks to touch taps for legacy website support. It may be possible to misuse drag events to simulate swiping with a mouse but that'll have to go in the backlog.

**Comment by Nerwyn** [comment]
Some significant refactors in the upcoming version 3.2.0 (now available as a beta) help pave the way for this long awaited feature request, although I'll have to work on porting the changes to my other project along with other features and bugs on those projects before coming back to this. Taking more notes:

> 1. One finger double tap
✅  Will work on all platforms (including iPhone) in v3.2.0 for buttons and touchpad.

> 2. Two finger swipes
🔜 Should be able to implement using [TouchEvent touches or targetTouches](https://developer.mozilla.org/en-US/docs/Web/API/TouchEvent/targetTouches).

> 3. Two finger tap
🔜 Now that the card will use the correct mouse vs touch events based on if touch events are available in the browser document, this should be implementable with touches or targetTouches

> 4. Two finger double tap.
🔜 Same as above. Should be possible after the v3.2.0 refactor.

> 5. Fast two finger swipes for scrolling.
❌  Unfortunately still probably not possible.

> 6. Long press
✅ Will work on all platforms (including iPhone) in v3.2.0 for buttons and touchpad.

v3.2.0 adds Home Assistant action like syntax for custom action. Custom actions can have a `tap_action`, `hold_action`, or `double_tap_action`. I'm thinking that I can add a new action type like `multi_touch_action` which will be triggered if more than one finger is used on touchpad taps or swipes.

**Comment by houser42** [comment]
Hello.
Very cool that this might be happening.
I installed the beta and it seems to wok as before.
Is it already possible to test the new gestures? If so are there instructions? Many thanks.

**Comment by Nerwyn** [comment]
They're in the beta branch README and release description.

**Comment by Nerwyn** [comment]
I thought you said it was working like before? It sounds like a caching issue again. The compiled file name extension is being changed to `.min.js` so that may be causing it.

**Comment by houser42** [comment]
> I thought you said it was working like before? It sounds like a caching issue again. The compiled file name extension is being changed to `.min.js` so that may be causing it.

I was mistaken, the beta had not loaded properly the first time I tried it. Apologies.
Should I try something for the `.min.js` to take?

**Comment by Nerwyn** [comment]
It should be as simple as refreshing (F5) or hard refreshing (CTRL F5). On mobile you may have to clear the apps cache. I've tested this beta successfully on Chrome, Brave, and Firefox on desktop and the Android and iPhone apps.

**Comment by houser42** [comment]
Tried this again today and the new 3.2.1 version just worked and the errors I saw went away.
Everything is working here now.
It however looks like at least on my system the deprecation that is mentioned in the release docs is not active yet
and it looks like I need to still use those for it to work. The below config for Kodi works with both hold and double-tap for back. So very happy here ;)

```
type: custom:android-tv-card
keyboard_id: media_player.kodi
keyboard_mode: KODI
navigation_row: touchpad
touchpad_height: 310px
enable_double_click: true
double_click_keycode: back
long_click_keycode: menu
custom_keys:
  up:
    icon: mdi:chevron-up
    service: kodi.call_method
    service_data:
      entity_id: media_player.kodi
      method: Input.Up
  down:
    icon: mdi:chevron-down
    service: kodi.call_method
    service_data:
      entity_id: media_player.kodi
      method: Input.Down
  left:
    icon: mdi:chevron-left
    service: kodi.call_method
    service_data:
      entity_id: media_player.kodi
      method: Input.Left
  right:
    icon: mdi:chevron-right
    service: kodi.call_method
    service_data:
      entity_id: media_player.kodi
      method: Input.Right
  center:
    icon: mdi:kodi
    service: kodi.call_method
    service_data:
      entity_id: media_player.kodi
      method: Input.Select
  back:
    icon: mdi:kodi
    service: kodi.call_method
    service_data:
      entity_id: media_player.kodi
      method: Input.Back
  menu:
    service: kodi.call_method
    service_data:
      entity_id: media_player.kodi
      method: Input.ContextMenu
```

**Comment by houser42** [comment]
Dear Nerwyn,
I want to take a moment to extend a big thanks and show my appreciation for this code as it has made my transition to HA for HT use possible. No-look navigation with swipe gestures for HT remotes really is something that you can’t go back from once you’ve tried it. Many, many thanks again for listening and making this possible!
Happy Holidays and all variations on the theme.
Be safe, be well.

**Comment by Nerwyn** [comment]
Thanks man. You were one of my first open source supporters and your feedback has been invaluable in improving this project. I'm glad you enjoy using this project so much. While I didn't create the original touchpad code I do think I have put so much work into improving it that it's much more functional now.

> It however looks like at least on my main system which is an iPhone with iOS 17.2.1 the deprecation that is mentioned in your release docs is not active yet and it looks like I still need to use those for it to work.

That's weird. It works on my dev iPhone SE. Your updated config should look like this:

```yaml
type: custom:android-tv-card
keyboard_id: media_player.kodi
keyboard_mode: KODI
rows:
  - - touchpad
touchpad_style:
  height: 310px
custom_actions:
  up:
    tap_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.Up
  down:
    tap_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.Down
  left:
    tap_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.Left
  right:
    tap_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.Right
  center:
    tap_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.Select
    double_tap_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.Back
    hold_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.ContextMenu
```

Deprecation (as I recently learned and was calling soft deprecation) means that it still works but is discouraged from use in favor of newer methods. Obsoletion means that it no longer works. I don't plan to obsolete any of the deprecated configuration options, just scrub them from the README in favor of the newer methods. You should be able to continue using the old config options indefinitely, but I'd encourage using the new methods for newer remotes as newer features (like the multi touch touchpad actions) will use the newer configuration methods.

**Comment by houser42** [comment]
You learn something everyday. Thanks! That is not what I thought deprecated meant ;)
Yes, your tweaked code works here for Kodi without the deprecated lines. Thanks!
Apple TV still not quite there here. Below is what works 100% with the deprecated lines. Not sure what I am missing..
Tweak it if you have time for reference? If not Happy Holidays again regardless!

```
type: custom:android-tv-card
remote_id: remote.appletv1
navigation_row: touchpad
touchpad_height: 340px
enable_double_click: true
double_click_keycode: menu
long_click_keycode: top_menu
custom_actions:
  up:
    icon: mdi:chevron-up
    key: up
  down:
    icon: mdi:chevron-down
    key: down
  left:
    icon: mdi:chevron-left
    key: left
  right:
    icon: mdi:chevron-right
    key: right
  center:
    icon: mdi:checkbox-blank-circle
    key: select
  power:
    icon: mdi:power
    key: suspend
  play:
    icon: mdi:play
    key: play
  pause:
    icon: mdi:pause
    key: pause
  menu:
    icon: mdi:apple
    key: menu
  home:
    icon: mdi:apple
    key: home
  home_hold1:
    icon: mdi:apple
    key: home_hold
  top_menu:
    icon: mdi:apple
    key: top_menu
  previous1:
    icon: mdi:apple
    key: previous
```

**Comment by Nerwyn** [comment]
Try this:

```yaml
type: custom:android-tv-card
remote_id: remote.appletv1
rows:
  - - touchpad
touchpad_style:
  height: 340px
custom_actions:
  up:
    tap_action:
      action: key
      key: up
  down:
    tap_action:
      action: key
      key: down
  left:
    tap_action:
      action: key
      key: left
  right:
    tap_action:
      action: key
      key: right
  center:
    tap_action:
      action: key
      key: select
    double_tap_action:
      action: key
      key: menu
    hold_action:
      action: key
      key: top_menu
  power:
    icon: mdi:power
    tap_action:
      action: key
      key: suspend
    key: suspend
  play:
    icon: mdi:play
    key: play
  pause:
    icon: mdi:pause
    tap_action:
      action: key
      key: pause
  home:
    icon: mdi:apple
    tap_action:
      action: key
      key: home
    hold_action:
      action: key
      key: home_hold
```

**Comment by houser42** [comment]
Yup, working. Appreciated!
Be safe, be well

**Comment by Nerwyn** [comment]
@houser42 here's the first beta! A basic custom actions example is in the release notes. It has multi touch swipe, hold, and double tap actions working.

https://github.com/Nerwyn/android-tv-card/releases/tag/3.3.0-beta.001

**Comment by houser42** [comment]
Very cool!
Just a question on terminology:
The example code  says "multi_tap_action" is that what we would use for multi touch swipe?
MAny thanks

**Comment by Nerwyn** [comment]
Yup! I wanted to limit the number of new action types I was adding so `multi_tap_action` is for both center multi taps and for multi finger directional swipes.

**Comment by houser42** [comment]
Quick test gives an error. "Required key not provided" 
Thisexample  is to scroll in lists on Kodi.
Can you see anything simply and obviously wrong by any chance?

```
custom_actions:
  up:
    tap_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.Up
    multi_tap_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.ExecuteAction
        action: pageup
  down:
    tap_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.Down
    multi_tap_action:
      action: call-service
      service: kodi.call_method
      service_data:
        entity_id: media_player.kodi
        method: Input.ExecuteAction
        action: pagedown
```

**Comment by houser42** [comment]
Just wanted to say again that I appreciate this card a lot!
In short you have now made all the changes that I needed for my HT purposes. Fab!

A question while am at it: 
Would it be possible for you to implement a "repeat command on hold" for buttons in this card by any chance?
For things like volume on a TV or receiver. It is a missing feature in HA AFAIK, as there is no release event in HA buttons ATM, as discussed [here](https://community.home-assistant.io/t/repeat-command-on-hold-button-for-things-like-volume-control/607175). Just curious if you have some black magic for this as you  had for the other stuff in this card ;)

**Comment by Nerwyn** [comment]
This card currently has repeat on hold for certain buttons as defined [here](https://github.com/Nerwyn/android-tv-card/blob/main/src%2Fclasses%2Fremote-button.ts#L77), including volume up/down. I can look into making this configurable on a custom action basis.

**Comment by houser42** [comment]
Ok, that code is not something I would know how to use off the bat ...;)
Sure, if you can look into repeat on hold  for custom action buttons I would be all ears!
Be safe, be well

**Comment by Nerwyn** [comment]
@houser42 beta 4 lets you make any action (barring keyboard actions) repeatable, whether they are a button or touchpad. I've only done some initial testing but it should work. Set `hold_action` `action` to "repeat" and then a the `tap_action` will repeat when held. More info in the release notes.

https://github.com/Nerwyn/android-tv-card/releases/tag/3.3.0-beta.004

**Comment by houser42** [comment]
So I tried this a few ways without success so far.
Could you comment on the code below please?

This what I was exiting but hierarchy seems wrong for this context:

```
type: custom:android-tv-card
rows:
  - volume_up
custom_actions:
  volume_up:
    icon: mdi:volume-plus
    hold_action:
      action: call-service
      service: 
      - media_player.volume_up
target:
  entity_id: media_player.marantz_avr
entity: media_player.marantz_avr
button_style:
  '--size': 32px
  '--padding': 10px

```

This is a regular button that does what I want but without repeat obviously. 
```
show_name: true
show_icon: false
type: button
tap_action:
  action: call-service
  service: media_player.volume_up
  target:
    entity_id:
      - media_player.marantz_avr
entity: media_player.marantz_avr
name: VOL PLUS DEV
icon: mdi:volume-off
```

**Comment by Nerwyn** [comment]
You have to set `hold_tap_action` or `multi_hold_tap_action` as follows:

```yaml
hold_action:
  action: repeat
```

Volume up/down already repeat by default so you shouldn't have to change anything. Your target field is in the wrong block though. You also set the service call to `hold_action` but it should still be `tap_action`. Since the default `hold_action` `action` you don't have to change anything for it. You would have to set `hold_action `action` to repeat if the action didn't repeat by default, like shown below for `channel_up`.

```yaml
type: custom:android-tv-card
rows:
  - volume_up
  - channel_up
custom_actions:
  volume_up:
    icon: mdi:volume-plus
    tap_action:
      action: call-service
      service: media_player.volume_up
      target:
        entity_id: media_player.marantz_avr
  channel_up:
    icon: mdi:arrow-up
    tap_action:
      action: call-service
      service: media_player.channel_up
      target:
        entity_id: media_player.marantz_avr
    hold_action:
      action: repeat
button_style:
  '--size': 32px
  '--padding': 10px
```

**Comment by houser42** [comment]
Ahh yes now I see, yup that works too!
Kudos!
One more minor thing when you have time?
For reference. Below is button that works great with this card to scroll lists in Kodi with repeat.
I am using the volume up/down key for this. Is that a good way or should we use something else?

I am however unable to make the size of the icon, so the buttons in the grid I am using gets too big:
<img width="342" alt="Screenshot 2024-02-01 at 00 29 18" src="https://github.com/Nerwyn/android-tv-card/assets/15192440/71a28c02-1543-423e-b4a7-e412159d2970">
I want something like this if possible
<img width="296" alt="Screenshot 2024-02-01 at 00 29 45" src="https://github.com/Nerwyn/android-tv-card/assets/15192440/403c5aca-74ff-4284-b098-191e19e9f828">
:

I had a similar issue with the search and text buttons, so the code I am using at the bottom is from that, but it does not work here. Maybe you can have a look at that and comment at some point? Many thanks for al this!

```
type: custom:android-tv-card
rows:
  - volume_up
custom_actions:
  volume_up:
    icon: mdi:arrow-down-bold
    tap_action: null
    action: call-service
    service: kodi.call_method
    data:
      method: Input.ExecuteAction
      action: pagedown
      entity_id: media_player.kodi
    hold_action:
      action: repeat
button_style:
  '--size': 16px
  '--padding': 20px
```

**Comment by Nerwyn** [comment]
The entire card itself has a 12px padding around the edges, which is likely causing your issue. You can use card-mod to remove it. Also `--padding` is not a valid field for this card.

```yaml
type: custom:android-tv-card
rows:
  - volume_up
custom_actions:
  volume_up:
    icon: mdi:arrow-down-bold
    tap_action: null
    action: call-service
    service: kodi.call_method
    data:
      method: Input.ExecuteAction
      action: pagedown
      entity_id: media_player.kodi
    hold_action:
      action: repeat
button_style:
  '--size': 16px
card_mod:
  style: |
    ha-card {
      padding: 0px !important;
    }
```

**Comment by houser42** [comment]
All working...
I am running out of both compliments for your excellent rode, your swift smart and kind replies as well as things to ask for.
Everything I hoped for when I started this thread 6 months ago is now in there, and then some..
In short... You rock!

**Comment by Nerwyn** [comment]
@houser42 I did a refactor which broke things and just got everything working again, along with some bug fixes for scrolling while holding down repeated buttons. Can you try beta 7 RC 1 and make sure it works correctly for you?

https://github.com/Nerwyn/android-tv-card/releases/tag/3.3.0-beta.007

**Comment by Nerwyn** [comment]
@houser42 can you try beta 8 RC 3 and make sure it works correctly for you?

https://github.com/Nerwyn/android-tv-card/releases/tag/3.3.0-beta.008

**Comment by Nerwyn** [comment]
Okay this one should be good

https://github.com/Nerwyn/android-tv-card/releases/tag/3.3.0-beta.009

**Comment by houser42** [comment]
Yes, I tested 3.3.0-beta.009 and it appears to work as expected here.

**Comment by Nerwyn** [comment]
Released version 3.3.0! 🎉 
https://github.com/Nerwyn/android-tv-card/releases/tag/3.3.0

**Comment by houser42** [comment]
Just a non-critical related concept question when you have a moment? As the title of this thread says "alll platforms" ;)

I am also using the HA desktop client in parallel with my main iOS client for some things but that GUI uses buttons rather than swipe gestures. Is it possible at all to unify and get the same swipe functionality for the desktop client? 
I would assume not, as a touch GUI is a fundamentally different thing. Just curious.  Many thanks.
